nasa/ow_simulator
# OCEANWATER-/822/824: Revise deliver sample movement to maintain samples + Halt arm movement on error

## Linked Issues:
| EPIC ⚡ | [OCEANWATER-680 / Sample Acquistion](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-680)<br>[OCEANWATER-551 / Fault Injection & Detection](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-551)  |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-822 / Arm movement should halt when arm fault when using rosservice interface](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-822) |
| Jira Ticket 🎟️ | [OCEANWATER-824 / Revise deliver sample movement such that it safely delivers samples into the bin](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-824) |
| Github :octocat:  | N/A |


## Summary of Changes
* Revise deliver sample movement such that it safely delivers samples into the bin
* Arm movement is halted when using the rosservice interface once an arm faults is reported

## Test Sample Delivery
* Launch the simulation using the Atacama Y1A terrain
```bash
roslaunch ow atacama_y1a.launch sim_regolith:=true
```
* Perform a grind operation followed by dig linear operation using default settings
```bash
python3 grind_action_client.py
python3 dig_linear_action_client.py
```
* Call the **DeliverSample** routine using the ROS Service interface
```bash
rosservice call /arm/deliver_sample # note this service no longer takes arguments
```
* Once the **DeliverSample** is complete most of the sample should reside in the **Sample Bin** as shown in the following video:

https://user-images.githubusercontent.com/606033/142026526-d7de6bb9-6dc3-4b00-8af9-1a2c600342ce.mp4


## Test Arm Movement Stops on Arm Faults while using the ROS Service Interface 
* Launch the simulation using the Atacama Y1A terrain
```bash
roslaunch ow atacama_y1a.launch
```
* Invoke any arm operation using the ros service interface, for example:
```bash
rosservice call /arm/unstow # or /arm/stow
```
* Enable a fault on any of the arm joints, the arm should stop movement immediately and stops responding until the fault has been cleared
